### PR TITLE
Fix: Respect required=False in add_lightning_class_args when subclass_mode=False

### LINKED ISSUE #20851: `add_lightning_class_args` `required` argument ignored if not using subclass mode
### Bug description

`LightningCLI`'s method `add_lightning_class_args` has an optional argument `required` which defaults to `True`. According to the documentation, when `required=True`, the user must specify the callback/class (or it must be default-constructable); when `required=False` it need not be. However, [the `required` argument is ignored when  `subclass_mode=False`](https://github.com/Lightning-AI/pytorch-lightning/blob/dd2912a045f1a1a0c87528b41c9854fc7454d4b0/src/lightning/pytorch/cli.py#L150
) as is the default. Simple example to reproduce:

```python
from lightning.pytorch import callbacks, cli

class FooCheckpoint(callbacks.ModelCheckpoint):

    # Makes one argument required so it can't be default-constructed.
    def __init__(self, dirpath, *args, **kwargs):
        super().__init__(dirpath, *args, **kwargs)


class FooCLI(cli.LightningCLI):
    def add_arguments_to_parser(
        self, parser: cli.LightningArgumentParser
    ) -> None:
        parser.add_lightning_class_args(
            FooCheckpoint,
            "checkpoint",
            required=False,
        )
```

Attempting to instantiate `FooCLI` with any choice of model and datamodule but without specify the checkpoint's required argument will result in a configuration error:

```
Validation failed: Key "fit.checkpoint.dirpath" is required but not included in config object or its value is None.
```

I don't have the full context but the fix may be non-trivial in that when `subclass_mode=False`, [`LightningCLI` calls `add_class_arguments` from the superclass](https://github.com/Lightning-AI/pytorch-lightning/blob/dd2912a045f1a1a0c87528b41c9854fc7454d4b0/src/lightning/pytorch/cli.py#L150), namely `jsonargparse.ArgumentParser`, and [that does not seem to support a `required` argument](https://github.com/omni-us/jsonargparse/blob/dcd0a5ae70d0ef5c7c053436375bcad9aa462c84/jsonargparse/_signatures.py#L46). 

At the very least though the documentation should probably be modified to explain that `required=True` is ignored unless `subclass_mode=True`.

### What version are you seeing the problem on?

v2.5

### Reproduced in studio

_No response_

### How to reproduce the bug

```python

```

### Error messages and logs

```
Validation failed: Key "fit.checkpoint.dirpath" is required but not included in config object or its value is None.
```

### Environment

<details>
  <summary>Current environment</summary>

* CUDA:
        - GPU:
                - NVIDIA GeForce GTX 1660 Ti
        - available:         True
        - version:           12.4
* Lightning:
        - lightning:         2.5.0.post0
        - lightning-cloud:   0.5.70
        - lightning-utilities: 0.14.3
        - pytorch-lightning: 2.5.0.post0
        - torch:             2.5.1
        - torchmetrics:      1.7.1
* Packages:
        - accelerate:        1.7.0
        - aiohappyeyeballs:  2.6.1
        - aiohttp:           3.11.18
        - aiosignal:         1.3.2
        - anyio:             4.9.0
        - arrow:             1.3.0
        - asttokens:         3.0.0
        - attrs:             25.3.0
        - autocommand:       2.2.2
        - backports.tarfile: 1.2.0
        - beautifulsoup4:    4.13.4
        - black:             24.10.0
        - blessed:           1.21.0
        - boto3:             1.38.18
        - botocore:          1.38.18
        - certifi:           2025.4.26
        - charset-normalizer: 3.4.2
        - click:             8.2.0
        - croniter:          1.3.15
        - dateutils:         0.6.12
        - decorator:         5.1.1
        - deepdiff:          7.0.1
        - deprecated:        1.2.13
        - docker-pycreds:    0.4.0
        - docstring-parser:  0.16
        - editor:            1.6.6
        - executing:         0.8.3
        - fastapi:           0.88.0
        - filelock:          3.18.0
        - flake8:            7.1.1
        - frozenlist:        1.6.0
        - fsspec:            2023.12.2
        - gitdb:             4.0.12
        - gitpython:         3.1.44
        - h11:               0.16.0
        - hf-xet:            1.1.2
        - huggingface-hub:   0.31.2
        - idna:              3.10
        - importlib-metadata: 8.5.0
        - importlib-resources: 6.5.2
        - inflect:           7.3.1
        - inquirer:          3.4.0
        - ipython:           9.1.0
        - ipython-pygments-lexers: 1.1.1
        - itsdangerous:      2.2.0
        - jaraco.collections: 5.1.0
        - jaraco.context:    5.3.0
        - jaraco.functools:  4.0.1
        - jaraco.text:       3.12.1
        - jedi:              0.19.2
        - jinja2:            3.1.6
        - jmespath:          1.0.1
        - jsonargparse:      4.40.0
        - lightning:         2.5.0.post0
        - lightning-cloud:   0.5.70
        - lightning-utilities: 0.14.3
        - lit:               17.0.6.dev0
        - markdown-it-py:    3.0.0
        - markupsafe:        3.0.2
        - matplotlib-inline: 0.1.6
        - maxwell:           0.2.6
        - mccabe:            0.7.0
        - mdurl:             0.1.2
        - mkl-service:       2.4.0
        - more-itertools:    10.3.0
        - mpmath:            1.3.0
        - multidict:         6.4.3
        - mypy-extensions:   1.0.0
        - networkx:          3.4.2
        - numpy:             2.2.6
        - nvidia-cublas-cu12: 12.4.5.8
        - nvidia-cuda-cupti-cu12: 12.4.127
        - nvidia-cuda-nvrtc-cu12: 12.4.127
        - nvidia-cuda-runtime-cu12: 12.4.127
        - nvidia-cudnn-cu12: 9.1.0.70
        - nvidia-cufft-cu12: 11.2.1.3
        - nvidia-curand-cu12: 10.3.5.147
        - nvidia-cusolver-cu12: 11.6.1.9
        - nvidia-cusparse-cu12: 12.3.1.170
        - nvidia-nccl-cu12:  2.21.5
        - nvidia-nvjitlink-cu12: 12.4.127
        - nvidia-nvtx-cu12:  12.4.127
        - opentelemetry-api: 1.30.0
        - ordered-set:       4.1.0
        - packaging:         24.2
        - parso:             0.8.4
        - pathspec:          0.10.3
        - pexpect:           4.8.0
        - pip:               25.1
        - platformdirs:      4.3.8
        - prompt-toolkit:    3.0.43
        - propcache:         0.3.1
        - protobuf:          5.29.4
        - psutil:            6.1.1
        - ptyprocess:        0.7.0
        - pure-eval:         0.2.2
        - pycodestyle:       2.12.1
        - pydantic:          1.10.22
        - pyflakes:          3.2.0
        - pygments:          2.19.1
        - pyjwt:             2.10.1
        - python-dateutil:   2.9.0.post0
        - python-multipart:  0.0.20
        - pytorch-lightning: 2.5.0.post0
        - pytz:              2025.2
        - pyyaml:            6.0.2
        - readchar:          4.2.1
        - regex:             2024.11.6
        - requests:          2.32.3
        - rich:              14.0.0
        - runs:              1.2.2
        - s3transfer:        0.12.0
        - safetensors:       0.5.3
        - sentry-sdk:        2.28.0
        - setproctitle:      1.3.6
        - setuptools:        78.1.1
        - six:               1.17.0
        - smmap:             5.0.2
        - sniffio:           1.3.1
        - soupsieve:         2.7
        - stack-data:        0.2.0
        - starlette:         0.22.0
        - starsessions:      1.3.0
        - sympy:             1.13.1
        - tokenizers:        0.21.1
        - tomli:             2.0.1
        - torch:             2.5.1
        - torchmetrics:      1.7.1
        - tqdm:              4.67.1
        - traitlets:         5.14.3
        - transformers:      4.51.3
        - triton:            3.1.0
        - typeguard:         4.3.0
        - types-python-dateutil: 2.9.0.20250516
        - typeshed-client:   2.7.0
        - typing-extensions: 4.12.2
        - urllib3:           2.4.0
        - uvicorn:           0.34.2
        - wandb:             0.18.7
        - wcwidth:           0.2.13
        - websocket-client:  1.8.0
        - websockets:        11.0.3
        - wheel:             0.45.1
        - wrapt:             1.17.0
        - xmod:              1.8.1
        - yarl:              1.20.0
        - yoyodyne:          0.3.2
        - yoyodyne-pretrained: 0.1.0
        - zipp:              3.21.0
* System:
        - OS:                Linux
        - architecture:
                - 64bit
                - ELF
        - processor:         x86_64
        - python:            3.13.2
        - release:           5.15.167.4-microsoft-standard-WSL2
        - version:           #1 SMP Tue Nov 5 00:21:55 UTC 2024

</details>

### More info

_No response_

### PR DESCRIPTION
**Pull request by KAVYANSHTYAGI**
What does this PR do?

This PR fixes an inconsistency in LightningCLI.add_lightning_class_args() where the required=False flag was being ignored when subclass_mode=False (which is the default).
According to the documentation, setting required=False should make providing the corresponding config optional — but in practice, argument registration was enforced regardless, leading to errors if required constructor parameters (e.g., dirpath in ModelCheckpoint) were not provided.

This change adds a conditional check that skips argument registration when:

    subclass_mode=False

    required=False

    and the config key (nested_key) is not provided by the user

Additionally, a test case was added to verify that this behavior works as expected.

Fixes #20851 https://github.com/Lightning-AI/pytorch-lightning/issues/20851
<details> <summary><b>Before submitting</b></summary>

Was this discussed/agreed via a GitHub issue? yes

Did you read the contributor guideline, Pull Request section? yes

Did you make sure your PR does only one thing, instead of bundling different changes together? yes

Did you write any new necessary tests? yes

Did you verify new and existing tests pass locally with your changes? yes

Did you update the documentation if necessary?

Did you list all the breaking changes introduced by this pull request?

Did you update the CHANGELOG? (not for typos, docs, test updates, or minor internal changes/refactors)

</details>
PR review

Anyone in the community is welcome to review the PR.

</details>
  Yes :)

### CODE REVIEW

**Comment by Borda** [comment]
@mauvilsa mind have look, pls

File: src/lightning/pytorch/cli.py [hunk 1]
```diff
@@ -145,15 +145,26 @@ def add_lightning_class_args(
...
             if issubclass(lightning_class, Callback):
                 self.callback_keys.append(nested_key)
+
+            # NEW LOGIC: If subclass_mode=False and required=False, only add if config provides this key
+            if not subclass_mode and not required:
+                config_path = f"{self.subcommand}.{nested_key}" if getattr(self, "subcommand", None) else nested_key
+                config = getattr(self, "config", {})
+                if not any(k.startswith(config_path) for k in config):
+                    # Skip adding class arguments
+                    return []
```
Comment (mauvilsa) [inline src/lightning/pytorch/cli.py:155]: This method is only called on parser construction, not during parsing. So at this point there isn't any config. Because of the `getattr(self, "config", {})` this doesn't fail but it will never do anything.

File: tests/tests_pytorch/test_cli.py [hunk 2]
```diff
@@ -1860,6 +1860,34 @@ def test_lightning_cli_args_and_sys_argv_warning():
         LightningCLI(TestModel, run=False, args=["--model.foo=789"])
 
 
+def test_add_class_args_required_false_skips_addition(tmp_path):
```
Comment (mauvilsa) [inline tests/tests_pytorch/test_cli.py:1863]: @KAVYANSHTYAGI you did not run the test to check whether the code was working. I know this because it is not possible that it works. When you open a pull request it would be good that beforehand you run the tests. And the test should fail without the code changes and succeed with them.

File: tests/tests_pytorch/test_cli.py [hunk 3]
```diff
@@ -1860,6 +1860,34 @@ def test_lightning_cli_args_and_sys_argv_warning():
...
+    from lightning.pytorch import callbacks, cli
+
+    class FooCheckpoint(callbacks.ModelCheckpoint):
+        def __init__(self, dirpath, *args, **kwargs):
+            super().__init__(dirpath, *args, **kwargs)
```
Comment (mauvilsa) [inline tests/tests_pytorch/test_cli.py:1868]:
```suggestion
```

There is no reason for this class since `ModelCheckpoint` could be used directly.

File: tests/tests_pytorch/test_cli.py [hunk 4]
```diff
@@ -1860,6 +1860,34 @@ def test_lightning_cli_args_and_sys_argv_warning():
...
+
+    class SimpleModel:
+        def __init__(self):
+            pass
+
+    class SimpleDataModule:
+        def __init__(self):
+            pass
```
Comment (mauvilsa) [inline tests/tests_pytorch/test_cli.py:1876]:
```suggestion
```
`LightningCLI` requires modules to inherit from the lightning classes. These don't so there is no way this would work. Also there is no need to implement new classes, that is what `BoringModel` and `BoringDataModule` are for.

File: tests/tests_pytorch/test_cli.py [hunk 5]
```diff
@@ -1860,6 +1860,34 @@ def test_lightning_cli_args_and_sys_argv_warning():
...
+
+    class FooCLI(cli.LightningCLI):
+        def __init__(self):
+            super().__init__(
+                model_class=SimpleModel, datamodule_class=SimpleDataModule, run=False, save_config_callback=None
+            )
```
Comment (mauvilsa) [inline tests/tests_pytorch/test_cli.py:1882]:
```suggestion
```
What is the point of this? It doesn't seem relevant to the test.

File: tests/tests_pytorch/test_cli.py [hunk 6]
```diff
@@ -1860,6 +1860,34 @@ def test_lightning_cli_args_and_sys_argv_warning():
...
+
+        def add_arguments_to_parser(self, parser):
+            parser.add_lightning_class_args(FooCheckpoint, "checkpoint", required=False)
```
Comment (mauvilsa) [inline tests/tests_pytorch/test_cli.py:1885]:
```suggestion
            parser.add_lightning_class_args(ModelCheckpoint, "checkpoint", required=False)
```

File: tests/tests_pytorch/test_cli.py [hunk 7]
```diff
@@ -1860,6 +1860,34 @@ def test_lightning_cli_args_and_sys_argv_warning():
...
+
+    # Expectation: No error raised even though FooCheckpoint requires `dirpath`
+    FooCLI()
```
Comment (mauvilsa) [inline tests/tests_pytorch/test_cli.py:1888]:
```suggestion
    FooCLI(BoringModel, BoringDataModule, run=False, save_config_callback=None, args=[])
```
This must have at least `args=[]`, otherwise it gets the arguments from the pytest call which for sure will not be valid for the CLI.

But anyway, `callback` + `required=False` should not be supported. But this test could be changed to assert that a specific error message is shown.

**Review by mauvilsa — CHANGES_REQUESTED** [review]
@Borda this should **not** be merged.

@KAVYANSHTYAGI unfortunately the presence of the `required` parameters mislead you. The only reason why `add_lightning_class_args` supports `Callback` is to add callbacks that must always be present, see [configure-forced-callbacks](https://lightning.ai/docs/pytorch/stable/cli/lightning_cli_expert.html#configure-forced-callbacks) in the docs. So simply `Callback` + `required=False` should give an error. But you are welcome to contribute that error logic instead.
